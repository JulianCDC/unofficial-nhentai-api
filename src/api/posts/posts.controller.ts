import {
  CacheInterceptor,
  CacheTTL,
  Controller,
  Get,
  Param,
  Query,
  Req,
  Res,
  UseInterceptors,
} from '@nestjs/common';
import {
  ApiBadRequestResponse,
  ApiNotFoundResponse,
  ApiOkResponse,
  ApiOperation,
  ApiProduces,
  ApiResponse,
  ApiTags,
} from '@nestjs/swagger';
import { Request, Response } from 'express';
import { PostsService } from './posts.service';
import { DownloadPostDto } from './types/download-post.dto';
import { GetDetailsDto } from './types/get-details.dto';
import { GetPostPageDto } from './types/get-post-page.dto';
import { GetPostDto } from './types/get-post.dto';
import { PostDetailsEntity } from './types/post-details.entity';

@ApiTags('galleries')
@Controller()
@UseInterceptors(CacheInterceptor)
export class PostsController {
  constructor(private readonly postsService: PostsService) {}

  @ApiOperation({
    summary: 'get details for a gallery',
  })
  @ApiBadRequestResponse()
  @ApiNotFoundResponse({ description: 'no gallery with provided id' })
  @Get('g/:id')
  async getGalleryById(
    @Param() params: GetPostDto,
    @Query() query: GetDetailsDto,
  ): Promise<PostDetailsEntity> {
    const details = await this.postsService.details(params.id);

    if (query.filters.length) {
      const filteredKeys = Object.keys(details).filter(key =>
        new RegExp(`(,|^)${key}(,|$)`, 'gi').test(query.filters),
      );

      const filteredDetails = {};
      filteredKeys.forEach(key => {
        filteredDetails[key] = details[key];
      });

      return filteredDetails;
    } else {
      return details;
    }
  }

  @ApiOperation({
    summary: 'get page of a gallery in jpeg format',
  })
  @ApiBadRequestResponse()
  @ApiNotFoundResponse({ description: 'page does not exist in gallery' })
  @ApiOkResponse({ schema: { type: 'string', format: 'binary' } })
  @ApiProduces('image/jpeg')
  @Get('g/:id/page/:page')
  async getGalleryPage(
    @Res() res: Response,
    @Param() params: GetPostPageDto,
  ): Promise<void> {
    const image = await this.postsService.page(params.id, params.page);

    res.setHeader('Content-Type', 'image/jpeg');
    image.pipe(res);
  }

  @ApiOperation({
    summary: 'get thumbnail for a gallery in jpeg format',
  })
  @ApiBadRequestResponse()
  @ApiNotFoundResponse({ description: 'no gallery with provided id' })
  @ApiOkResponse({ schema: { type: 'string', format: 'binary' } })
  @ApiProduces('image/jpeg')
  @Get('g/:id/thumbnail')
  async getGalleryThumbnail(
    @Res() res: Response,
    @Param() params: GetPostDto,
  ): Promise<void> {
    const image = await this.postsService.thumbnail(params.id);

    res.setHeader('Content-Type', 'image/jpeg');
    image.pipe(res);
  }

  @ApiOperation({
    summary: 'download specified number of pages for a gallery in a zip file',
  })
  @ApiBadRequestResponse()
  @ApiNotFoundResponse({ description: 'no gallery with provided id' })
  @ApiOkResponse({ schema: { type: 'string', format: 'binary' } })
  @ApiProduces('application/zip')
  @Get('g/:id/download')
  async getGalleryAsZip(
    @Res() res: Response,
    @Param() params: GetPostDto,
    @Query() query: DownloadPostDto,
  ): Promise<void> {
    res.setHeader('Content-Type', 'application/zip');
    res.setHeader(
      'Content-Disposition',
      `attachment; filename=${params.id}.zip`,
    );

    this.postsService.download(res, params.id, query.numberOfPages);
  }

  @ApiResponse({
    status: 303,
    description: 'redirect to /g/:id with random post id',
  })
  @ApiOperation({
    summary: 'get a random gallery',
    description:
      'can be chained with other routes (example: /posts/random/thumbnail',
  })
  @Get('posts/random(/*)?')
  @CacheTTL(1)
  async getRandomGallery(
    @Req() req: Request,
    @Res() res: Response,
  ): Promise<void> {
    const randomId = await this.postsService.random();

    const ressource = req.url.split('random/')[1] ?? '';

    res.setHeader('Cache-Control', 'no-cache');

    res.redirect(`/g/${randomId}/${ressource}`);
  }
}
