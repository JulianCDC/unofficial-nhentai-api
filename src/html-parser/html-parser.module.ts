import { Module, Global } from '@nestjs/common';
import { HtmlParserService } from './html-parser/html-parser.service';

@Global()
@Module({
  providers: [HtmlParserService],
  exports: [HtmlParserService],
})
export class HtmlParserModule {}
