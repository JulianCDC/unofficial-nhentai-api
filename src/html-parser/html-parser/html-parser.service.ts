import { Injectable } from '@nestjs/common';
import fetch from 'node-fetch';
import { load } from 'cheerio';

@Injectable()
export class HtmlParserService {
  async parse(url: string): Promise<CheerioStatic> {
    const html = await this.fetchPage(url);
    const $ = load(html);

    return $;
  }

  async mapParse<T>(
    url: string,
    selector: CheerioSelector | string,
    map: (element: Cheerio) => T,
  ): Promise<T[]> {
    const $ = await this.parse(url);

    return $(selector)
      .toArray()
      .map(element => map($(element)));
  }

  private async fetchPage(url: string): Promise<string> {
    return fetch(url).then(response => response.text());
  }
}
