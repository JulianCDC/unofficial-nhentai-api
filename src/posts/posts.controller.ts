import {
  Controller,
  Get,
  Param,
  Query,
  UseInterceptors,
  CacheInterceptor,
} from '@nestjs/common';
import { PostsService } from './posts.service';
import { DetailsQuery } from './types/details-query.type';
import { PostParam } from './types/post-param.type';
import { DetailsResponse } from './types/details-response.type';

@Controller('g')
@UseInterceptors(CacheInterceptor)
export class PostsController {
  constructor(private readonly postsService: PostsService) {}

  @Get(':id')
  async pages(
    @Param() params: PostParam,
    @Query() query: DetailsQuery,
  ): Promise<DetailsResponse> {
    const details = await this.postsService.details(params.id);

    if (query.filters.length) {
      const filteredKeys = Object.keys(details).filter(key =>
        query.filters.includes(key.toLowerCase()),
      );

      const filteredDetails = {};
      filteredKeys.forEach(key => {
        filteredDetails[key] = details[key];
      });

      return filteredDetails;
    } else {
      return details;
    }
  }
}
