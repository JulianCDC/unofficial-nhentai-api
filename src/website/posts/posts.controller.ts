import {
  Controller,
  Get,
  Param,
  Query,
  UseInterceptors,
  CacheInterceptor,
  Res,
  Redirect,
  CacheTTL,
} from '@nestjs/common';
import { PostsService } from './posts.service';
import { GetDetailsDto } from './types/get-details.dto';
import { GetPostDto } from './types/get-post.dto';
import { PostDetailsEntity } from './types/post-details.entity';
import { Response } from 'express';
import { GetPostPageDto } from './types/get-post-page.dto';
import { DownloadPostDto } from './types/download-post.dto';
import {
  ApiBadRequestResponse,
  ApiMovedPermanentlyResponse,
  ApiNotFoundResponse,
  ApiOkResponse,
  ApiProduces,
  ApiResponse,
  ApiTags,
} from '@nestjs/swagger';

@ApiTags('galleries')
@Controller()
@UseInterceptors(CacheInterceptor)
export class PostsController {
  constructor(private readonly postsService: PostsService) {}

  @ApiBadRequestResponse()
  @ApiNotFoundResponse({ description: 'no gallery with provided id' })
  @Get('g/:id')
  async details(
    @Param() params: GetPostDto,
    @Query() query: GetDetailsDto,
  ): Promise<PostDetailsEntity> {
    const details = await this.postsService.details(params.id);

    if (query.filters.length) {
      // TODO: Strict check
      const filteredKeys = Object.keys(details).filter(key =>
        query.filters.includes(key.toLowerCase()),
      );

      const filteredDetails = {};
      filteredKeys.forEach(key => {
        filteredDetails[key] = details[key];
      });

      return filteredDetails;
    } else {
      return details;
    }
  }

  @ApiBadRequestResponse()
  @ApiNotFoundResponse({ description: 'page does not exist in gallery' })
  @ApiOkResponse({ schema: { type: 'string', format: 'binary' } })
  @ApiProduces('image/jpeg')
  @Get('g/:id/page/:page')
  async page(
    @Res() res: Response,
    @Param() params: GetPostPageDto,
  ): Promise<void> {
    const image = await this.postsService.page(params.id, params.page);

    res.setHeader('Content-Type', 'image/jpeg');
    image.pipe(res);
  }

  @ApiBadRequestResponse()
  @ApiNotFoundResponse({ description: 'no gallery with provided id' })
  @ApiOkResponse({ schema: { type: 'string', format: 'binary' } })
  @ApiProduces('image/jpeg')
  @Get('g/:id/thumbnail')
  async thumbnail(
    @Res() res: Response,
    @Param() params: GetPostDto,
  ): Promise<void> {
    const image = await this.postsService.thumbnail(params.id);

    res.setHeader('Content-Type', 'image/jpeg');
    image.pipe(res);
  }

  @ApiBadRequestResponse()
  @ApiNotFoundResponse({ description: 'no gallery with provided id' })
  @ApiOkResponse({ schema: { type: 'string', format: 'binary' } })
  @ApiProduces('application/zip')
  @Get('g/:id/download')
  async download(
    @Res() res: Response,
    @Param() params: GetPostDto,
    @Query() query: DownloadPostDto,
  ): Promise<void> {
    res.setHeader('Content-Type', 'application/zip');

    this.postsService.download(res, params.id, query.numberOfPages);
  }

  @ApiResponse({
    status: 303,
    description: 'redirect to /g/:id with random post id',
  })
  @Get('posts/random')
  @CacheTTL(1)
  @Redirect('/g', 303)
  async random(): Promise<{ url: string }> {
    const randomId = await this.postsService.random();

    return {
      url: `/g/${randomId}`,
    };
  }
}
