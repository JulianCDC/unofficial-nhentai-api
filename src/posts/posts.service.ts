import { Injectable, NotFoundException } from '@nestjs/common';
import { HtmlParserService } from '../html-parser/html-parser/html-parser.service';
import { DetailsResponse } from './types/details-response.type';
import { parseTags } from '../common/helpers/parse-tags.helper';
import fetch from 'node-fetch';

@Injectable()
export class PostsService {
  constructor(private readonly htmlParser: HtmlParserService) {}

  async details(id: number): Promise<Required<DetailsResponse>> {
    const $ = await this.htmlParser.parse(`https://nhentai.net/g/${id}/`);

    if (!$('.title').text()) {
      throw new NotFoundException();
    }

    const details: Required<DetailsResponse> = {
      id: id,
      name: {
        before: $('#info > h1 > span.before')
          .text()
          .trim(),
        after: $('#info > h1 > span.after')
          .text()
          .trim(),
        content: $('#info > h1 > span.pretty')
          .text()
          .trim(),
      },
      thumbnail: $('#cover > a > img').data('src'),
      pages: parseInt($('#tags > div:nth-child(8) > span > a > span').text()),
      uploadDate: $('.tags > time').attr('datetime'),
      parodies: parseTags($('#tags > div:nth-child(1) > span > .tag'), $),
      characters: parseTags($('#tags > div:nth-child(2) > span > .tag'), $),
      tags: parseTags($('#tags > div:nth-child(3) > span > .tag'), $),
      artists: parseTags($('#tags > div:nth-child(4) > span > .tag'), $),
      groups: parseTags($('#tags > div:nth-child(5) > span > .tag'), $),
      languages: parseTags($('#tags > div:nth-child(6) > span > .tag'), $),
      categories: parseTags($('#tags > div:nth-child(7) > span > .tag'), $),
    };

    return details;
  }

  async page(id: number, page: number): Promise<NodeJS.ReadableStream> {
    const $ = await this.htmlParser.parse(
      `https://nhentai.net/g/${id}/${page}`,
    );

    const imageURI = $('#image-container > a > img').attr('src');

    if (!imageURI) {
      throw new NotFoundException();
    }

    const image = await fetch(imageURI);

    return image.body;
  }

  async thumbnail(id: number): Promise<NodeJS.ReadableStream> {
    const $ = await this.htmlParser.parse(`https://nhentai.net/g/${id}`);

    const imageURI = $('#cover > a > img').attr('data-src');

    if (!imageURI) {
      throw new NotFoundException();
    }

    console.log(imageURI);
    const image = await fetch(imageURI);

    return image.body;
  }

  async random(): Promise<number> {
    const randomPage = await fetch('https://nhentai.net/random');
    const uri = randomPage.url.split('/');
    const randomId = uri[uri.length - 2];

    return parseInt(randomId);
  }
}
