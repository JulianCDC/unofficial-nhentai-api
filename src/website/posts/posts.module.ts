import { CacheModule, Module } from '@nestjs/common';
import { HtmlParserModule } from '../../html-parser/html-parser.module';
import { PostsController } from './posts.controller';
import { PostsService } from './posts.service';

@Module({
  imports: [HtmlParserModule, CacheModule.register({ ttl: 3600 })],
  controllers: [PostsController],
  providers: [PostsService],
})
export class PostsModule {}
