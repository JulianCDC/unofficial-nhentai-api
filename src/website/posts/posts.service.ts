import { Injectable, NotFoundException } from '@nestjs/common';
import { HtmlParserService } from '../../html-parser/html-parser/html-parser.service';
import { DetailsResponse } from './types/details-response.type';
import { parseTags } from '../common/helpers/parse-tags.helper';
import fetch from 'node-fetch';
import * as archiver from 'archiver';

@Injectable()
export class PostsService {
  constructor(private readonly htmlParser: HtmlParserService) {}

  async details(id: number): Promise<Required<DetailsResponse>> {
    const $ = await this.htmlParser.parse(`https://nhentai.net/g/${id}/`);

    if (!$('.title').text()) {
      throw new NotFoundException();
    }

    const details: Required<DetailsResponse> = {
      id: id,
      name: {
        before: $('#info > h1 > span.before')
          .text()
          .trim(),
        after: $('#info > h1 > span.after')
          .text()
          .trim(),
        content: $('#info > h1 > span.pretty')
          .text()
          .trim(),
      },
      thumbnail: $('#cover > a > img').data('src'),
      pages: parseInt($('#tags > div:nth-child(8) > span > a > span').text()),
      uploadDate: $('.tags > time').attr('datetime'),
      parodies: parseTags($('#tags > div:nth-child(1) > span > .tag'), $),
      characters: parseTags($('#tags > div:nth-child(2) > span > .tag'), $),
      tags: parseTags($('#tags > div:nth-child(3) > span > .tag'), $),
      artists: parseTags($('#tags > div:nth-child(4) > span > .tag'), $),
      groups: parseTags($('#tags > div:nth-child(5) > span > .tag'), $),
      languages: parseTags($('#tags > div:nth-child(6) > span > .tag'), $),
      categories: parseTags($('#tags > div:nth-child(7) > span > .tag'), $),
    };

    return details;
  }

  async page(id: number, page: number): Promise<NodeJS.ReadableStream> {
    const $ = await this.htmlParser.parse(
      `https://nhentai.net/g/${id}/${page}`,
    );

    const imageURI = $('#image-container > a > img').attr('src');

    if (!imageURI) {
      throw new NotFoundException();
    }

    const image = await fetch(imageURI);

    return image.body;
  }

  async thumbnail(id: number): Promise<NodeJS.ReadableStream> {
    const $ = await this.htmlParser.parse(`https://nhentai.net/g/${id}`);

    const imageURI = $('#cover > a > img').attr('data-src');

    if (!imageURI) {
      throw new NotFoundException();
    }

    console.log(imageURI);
    const image = await fetch(imageURI);

    return image.body;
  }

  async download(id: number, pages = 1): Promise<archiver.Archiver> {
    const archive = archiver('zip');

    for (let i = 1; i < pages; i++) {
      const image = await this.page(id, i);
      const buffer = [];

      await new Promise(resolve => {
        image.on('data', data => {
          buffer.push(data);
        });
        image.on('end', () => {
          resolve();
          archive.append(Buffer.concat(buffer), { name: `${i}.jpeg` });
        });
      });
    }

    archive.finalize();

    return archive;
  }

  async random(): Promise<number> {
    const randomPage = await fetch('https://nhentai.net/random');
    const uri = randomPage.url.split('/');
    const randomId = uri[uri.length - 2];

    return parseInt(randomId);
  }
}
