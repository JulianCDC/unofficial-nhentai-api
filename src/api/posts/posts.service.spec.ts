import { CacheModule, NotFoundException } from '@nestjs/common';
import { Test } from '@nestjs/testing';
import { load } from 'cheerio';
import { random } from 'faker';
import * as fs from 'fs';
import { FetchMock } from 'jest-fetch-mock';
import fetch from 'node-fetch';
import { HtmlParserModule } from 'src/html-parser/html-parser.module';
import { HtmlParserService } from 'src/html-parser/html-parser/html-parser.service';
import { PostsService } from './posts.service';

const fetchMock = (fetch as unknown) as FetchMock;

describe('CategoriesController', () => {
  let postsService: PostsService;
  let htmlParserService: HtmlParserService;
  let galleryPage: string;

  beforeAll(async () => {
    galleryPage = fs.readFileSync(
      'test/mocks/pages/page-with-tags.html',
      'utf8',
    );
  });

  beforeEach(async () => {
    const moduleRef = await Test.createTestingModule({
      controllers: [],
      providers: [PostsService],
      imports: [CacheModule.register({ ttl: 1 }), HtmlParserModule],
    }).compile();

    postsService = moduleRef.get(PostsService);
    htmlParserService = moduleRef.get(HtmlParserService);
  });

  describe('details', () => {
    it('Returns gallery details', async () => {
      htmlParserService.parse = jest
        .fn()
        .mockReturnValueOnce(load(galleryPage));

      await expect(postsService.details(177013)).resolves.toMatchSnapshot();
    });

    it('Throws error when gallery does not exist', async () => {
      htmlParserService.parse = jest
        .fn()
        .mockReturnValueOnce(
          load(fs.readFileSync('test/mocks/pages/not-found-page.html')),
        );

      await expect(postsService.details(0)).rejects.toThrowError(
        NotFoundException,
      );
    });
  });

  describe('random', () => {
    it.only('Returns random page', async () => {
      const randomId = random.number(999999);

      fetchMock.mockIf('https://nhentai.net/random', async () => {
        return {
          url: `https://nhentai.net/g/${randomId}/`,
        };
      });

      await expect(postsService.random()).resolves.toBe(randomId);
    });
  });
});
