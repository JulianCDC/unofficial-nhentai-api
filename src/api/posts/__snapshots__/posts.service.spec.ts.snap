// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoriesController details Returns gallery details 1`] = `
Object {
  "artists": Array [
    Object {
      "id": 3981,
      "name": "shindol",
      "tagged": 279,
      "uri": "/artist/shindol/",
    },
  ],
  "categories": Array [
    Object {
      "id": 33173,
      "name": "manga",
      "tagged": 77000,
      "uri": "/category/manga/",
    },
  ],
  "characters": Array [],
  "groups": Array [],
  "id": 177013,
  "languages": Array [
    Object {
      "id": 17249,
      "name": "translated",
      "tagged": 109000,
      "uri": "/language/translated/",
    },
    Object {
      "id": 12227,
      "name": "english",
      "tagged": 69000,
      "uri": "/language/english/",
    },
  ],
  "name": Object {
    "after": "(Complete) [English]",
    "before": "[ShindoLA]",
    "content": "METAMORPHOSIS",
  },
  "pages": 225,
  "parodies": Array [],
  "tags": Array [
    Object {
      "id": 8010,
      "name": "group",
      "tagged": 72000,
      "uri": "/tag/group/",
    },
    Object {
      "id": 24201,
      "name": "stockings",
      "tagged": 64000,
      "uri": "/tag/stockings/",
    },
    Object {
      "id": 14283,
      "name": "anal",
      "tagged": 64000,
      "uri": "/tag/anal/",
    },
    Object {
      "id": 10314,
      "name": "schoolgirl uniform",
      "tagged": 56000,
      "uri": "/tag/schoolgirl-uniform/",
    },
    Object {
      "id": 13720,
      "name": "nakadashi",
      "tagged": 42000,
      "uri": "/tag/nakadashi/",
    },
    Object {
      "id": 29859,
      "name": "blowjob",
      "tagged": 37000,
      "uri": "/tag/blowjob/",
    },
    Object {
      "id": 13989,
      "name": "ahegao",
      "tagged": 29000,
      "uri": "/tag/ahegao/",
    },
    Object {
      "id": 22942,
      "name": "incest",
      "tagged": 28000,
      "uri": "/tag/incest/",
    },
    Object {
      "id": 22945,
      "name": "double penetration",
      "tagged": 24000,
      "uri": "/tag/double-penetration/",
    },
    Object {
      "id": 19018,
      "name": "dark skin",
      "tagged": 21000,
      "uri": "/tag/dark-skin/",
    },
    Object {
      "id": 20035,
      "name": "x-ray",
      "tagged": 20000,
      "uri": "/tag/x-ray/",
    },
    Object {
      "id": 29224,
      "name": "impregnation",
      "tagged": 15000,
      "uri": "/tag/impregnation/",
    },
    Object {
      "id": 27384,
      "name": "mind break",
      "tagged": 13000,
      "uri": "/tag/mind-break/",
    },
    Object {
      "id": 7256,
      "name": "mmf threesome",
      "tagged": 9000,
      "uri": "/tag/mmf-threesome/",
    },
    Object {
      "id": 6343,
      "name": "pregnant",
      "tagged": 9000,
      "uri": "/tag/pregnant/",
    },
    Object {
      "id": 8739,
      "name": "story arc",
      "tagged": 9000,
      "uri": "/tag/story-arc/",
    },
    Object {
      "id": 22079,
      "name": "drugs",
      "tagged": 6000,
      "uri": "/tag/drugs/",
    },
    Object {
      "id": 12695,
      "name": "prostitution",
      "tagged": 5000,
      "uri": "/tag/prostitution/",
    },
    Object {
      "id": 5820,
      "name": "piercing",
      "tagged": 5000,
      "uri": "/tag/piercing/",
    },
    Object {
      "id": 29182,
      "name": "blackmail",
      "tagged": 4000,
      "uri": "/tag/blackmail/",
    },
    Object {
      "id": 25050,
      "name": "gyaru",
      "tagged": 3000,
      "uri": "/tag/gyaru/",
    },
    Object {
      "id": 32996,
      "name": "deepthroat",
      "tagged": 3000,
      "uri": "/tag/deepthroat/",
    },
    Object {
      "id": 10542,
      "name": "snuff",
      "tagged": 2000,
      "uri": "/tag/snuff/",
    },
    Object {
      "id": 53449,
      "name": "already uploaded",
      "tagged": 2000,
      "uri": "/tag/already-uploaded/",
    },
    Object {
      "id": 7288,
      "name": "vomit",
      "tagged": 878,
      "uri": "/tag/vomit/",
    },
    Object {
      "id": 21112,
      "name": "full body tattoo",
      "tagged": 586,
      "uri": "/tag/full-body-tattoo/",
    },
    Object {
      "id": 13722,
      "name": "moral degeneration",
      "tagged": 573,
      "uri": "/tag/moral-degeneration/",
    },
  ],
  "thumbnail": "https://t.nhentai.net/galleries/987560/cover.jpg",
  "uploadDate": "2016-10-18T12:28:49.949133+00:00",
}
`;
